net/ipsec/ipsec.go

```diff
@@ -0,0 +1,732 @@
+// package IPsec provides primitives for establishing IPsec in the fastdp mode.
+package ipsec
+
+// TODO:
+// * atomic inserts
+// * add logging here for better diagnostics
+//
+// * test with non-default ports
+// * test on larger cluster
+//
+// * user-configurable life-times
+// * tests for rekeying
+// * check flow
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"sync"
+	"syscall"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netlink/nl"
+	"golang.org/x/crypto/hkdf"
+
+	"github.com/weaveworks/mesh"
+)
+
+const (
+	protoVsn = 1 // fastdp-crypto vsn
+
+	keySize   = 36 // AES-GCM key 32 bytes + 4 bytes salt
+	nonceSize = 32 // HKDF nonce size
+
+	mark    = uint32(0x1) << 17 // iptables marks
+	markStr = "0x20000/0x20000" // update if mark changes
+
+	tableMangle  = "mangle"
+	tableFilter  = "filter"
+	chainIn      = "WEAVE-IPSEC-IN"
+	chainInMark  = "WEAVE-IPSEC-IN-MARK"
+	chainOut     = "WEAVE-IPSEC-OUT"
+	chainOutMark = "WEAVE-IPSEC-OUT-MARK"
+
+	limitByteHard = 350 * 1024 * 1024 // 350mb
+	//limitTimeHard = 30 * 60           // 30min
+	limitTimeHard = 60 // 1min (for testing)
+)
+
+type SPI uint32
+
+// Used to identify:
+// - directional SPIs,
+// - ipsec establishments.
+type spiID [16]byte
+
+type spiInfo struct {
+	spi       SPI
+	isDirOut  bool
+	initRekey func() error
+}
+
+func getSPIId(srcPeer, dstPeer mesh.PeerName) (id spiID) {
+	binary.BigEndian.PutUint64(id[:], uint64(srcPeer))
+	binary.BigEndian.PutUint64(id[8:], uint64(dstPeer))
+	return
+}
+
+// IPSec
+
+type IPSec struct {
+	sync.RWMutex
+	ipt *iptables.IPTables
+
+	rc      *connRefCount
+	spiInfo map[spiID]spiInfo
+	spis    map[SPI]*spiInfo
```

Comment:
This is a reference to `spiInfo` which might be expired, i.e. `spiInfo.spi` has been expired.